File: quick_accuracy_test.py
#!/usr/bin/env python3
"""Quick accuracy test - sample 200 colors evenly across the dataset"""

import csv
import subprocess
import numpy as np
from colour.notation import munsell
from colour import sRGB_to_XYZ, XYZ_to_xyY

# Load all colors
colors = []
with open('tests/data/srgb-to-munsell.csv', 'r') as f:
    reader = csv.reader(f)
    next(reader)  # Skip header
    for row in reader:
        colors.append((int(row[0]), int(row[1]), int(row[2]), row[3]))

# Sample evenly
sample_size = 200
indices = np.linspace(0, len(colors)-1, sample_size, dtype=int)
sampled = [colors[i] for i in indices]

print(f"Testing {sample_size} colors sampled from {len(colors)} total...")

stats = {
    'hue_diffs': [],
    'value_diffs': [],
    'chroma_diffs': [],
    'family_mismatches': 0,
    'total': 0,
    'within_tolerance': 0
}

for i, (r, g, b, expected) in enumerate(sampled):
    if i % 20 == 0:
        print(f"Progress: {i}/{sample_size}")
    
    # Python conversion
    try:
        srgb = [r/255.0, g/255.0, b/255.0]
        xyz = sRGB_to_XYZ(srgb)
        xyy = XYZ_to_xyY(xyz)
        py_spec = munsell.xyY_to_munsell_specification(xyy)
    except:
        continue
    
    # Rust conversion
    result = subprocess.run(
        ['./target/release/test_rgb_cli', str(r), str(g), str(b)],
        capture_output=True, text=True, timeout=1
    )
    
    if result.returncode != 0 or 'Munsell:' not in result.stdout:
        continue
    
    rust_notation = result.stdout.split('Munsell:')[1].strip()
    
    try:
        rust_spec = munsell.munsell_colour_to_munsell_specification(rust_notation)
    except:
        continue
    
    # Calculate differences
    h_diff = abs(py_spec[0] - rust_spec[0]) if not (np.isnan(py_spec[0]) or np.isnan(rust_spec[0])) else 0
    v_diff = abs(py_spec[1] - rust_spec[1])
    c_diff = abs(py_spec[2] - rust_spec[2]) if not (np.isnan(py_spec[2]) or np.isnan(rust_spec[2])) else 0
    
    stats['hue_diffs'].append(h_diff)
    stats['value_diffs'].append(v_diff)
    stats['chroma_diffs'].append(c_diff)
    stats['total'] += 1
    
    if h_diff <= 0.1 and v_diff <= 0.1 and c_diff <= 0.1:
        stats['within_tolerance'] += 1
    
    if int(py_spec[3]) != int(rust_spec[3]):
        stats['family_mismatches'] += 1

# Print results
print("\n" + "="*80)
print("QUICK ACCURACY TEST RESULTS")
print("="*80)

accuracy = 100 * stats['within_tolerance'] / stats['total'] if stats['total'] > 0 else 0
print(f"\nOverall Accuracy: {accuracy:.1f}% ({stats['within_tolerance']}/{stats['total']} within 0.1 tolerance)")
print(f"Family Mismatches: {stats['family_mismatches']} ({100*stats['family_mismatches']/stats['total']:.1f}%)")

print(f"\nComponent Statistics:")
for component, diffs in [('Hue', stats['hue_diffs']), 
                         ('Value', stats['value_diffs']), 
                         ('Chroma', stats['chroma_diffs'])]:
    if diffs:
        print(f"\n{component}:")
        print(f"  Median: {np.median(diffs):.6f}")
        print(f"  90th %ile: {np.percentile(diffs, 90):.6f}")
        print(f"  95th %ile: {np.percentile(diffs, 95):.6f}")
        print(f"  99th %ile: {np.percentile(diffs, 99):.6f}")
        print(f"  Max: {max(diffs):.6f}")
        print(f"  >0.1: {sum(1 for d in diffs if d > 0.1)} ({100*sum(1 for d in diffs if d > 0.1)/len(diffs):.1f}%)")

print("="*80)

File: quick_accuracy_test.rs
use munsellspace::mathematical::{
    MathematicalMunsellConverter, 
    Illuminant, 
    ChromaticAdaptation
};
use std::fs::File;
use std::io::{BufRead, BufReader};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("Quick accuracy test with restored breakthrough version\n");
    
    // Load reference dataset
    let file = File::open("tests/data/srgb-to-munsell.csv")?;
    let reader = BufReader::new(file);
    
    let mut reference_data = Vec::new();
    for line in reader.lines().skip(1) {
        let line = line?;
        let parts: Vec<&str> = line.split(',').collect();
        if parts.len() >= 4 {
            let r = parts[0].trim().parse::<u8>().unwrap_or(0);
            let g = parts[1].trim().parse::<u8>().unwrap_or(0);
            let b = parts[2].trim().parse::<u8>().unwrap_or(0);
            let munsell = parts[3].trim().to_string();
            reference_data.push(([r, g, b], munsell));
        }
    }
    
    println!("Loaded {} reference colors", reference_data.len());
    
    // Test default configuration (should maintain 60.4% accuracy)
    let converter = MathematicalMunsellConverter::new()?;
    
    // Sample every 10th color for speed
    let sample_size = 400;
    let step = reference_data.len() / sample_size;
    
    let mut exact_matches = 0;
    let mut family_matches = 0;
    let mut colors_tested = 0;
    
    for i in (0..reference_data.len()).step_by(step) {
        let (rgb, expected) = &reference_data[i];
        
        // Set a timeout environment variable for convergence
        std::env::set_var("MUNSELL_MAX_ITERATIONS", "10");
        
        let result = converter.srgb_to_munsell(*rgb)?;
        let notation = format!("{:.1}{} {:.1}/{:.1}", 
            result.hue, result.family, result.value, result.chroma);
        
        colors_tested += 1;
        
        if notation == *expected {
            exact_matches += 1;
        }
        
        // Extract family from expected
        if let Some(space_pos) = expected.find(' ') {
            let hue_part = &expected[..space_pos];
            let expected_family = hue_part.chars()
                .skip_while(|c| c.is_numeric() || *c == '.')
                .collect::<String>();
            
            if result.family == expected_family {
                family_matches += 1;
            }
        }
        
        // Print progress every 50 colors
        if colors_tested % 50 == 0 {
            print!(".");
            use std::io::{self, Write};
            io::stdout().flush().unwrap();
        }
    }
    println!();
    
    let exact_pct = (exact_matches as f64 / colors_tested as f64) * 100.0;
    let family_pct = (family_matches as f64 / colors_tested as f64) * 100.0;
    
    println!("\nColors tested: {} (sampled from {})", colors_tested, reference_data.len());
    println!("Exact matches: {}/{} ({:.1}%)", exact_matches, colors_tested, exact_pct);
    println!("Family matches: {}/{} ({:.1}%)", family_matches, colors_tested, family_pct);
    
    println!("\n=== Testing with different illuminants ===");
    
    // Test a few colors with different illuminants
    let test_colors = [
        ([255, 0, 0], "pure red"),
        ([0, 255, 0], "pure green"),
        ([0, 0, 255], "pure blue"),
        ([255, 255, 0], "yellow"),
        ([128, 128, 128], "gray"),
    ];
    
    for (rgb, name) in test_colors {
        println!("\n{} RGB{:?}:", name, rgb);
        
        // D65 (default)
        let d65_converter = MathematicalMunsellConverter::new()?;
        let d65_result = d65_converter.srgb_to_munsell(rgb)?;
        println!("  D65: {:.1}{} {:.1}/{:.1}", 
            d65_result.hue, d65_result.family, d65_result.value, d65_result.chroma);
        
        // Illuminant C with XYZScaling
        let c_converter = MathematicalMunsellConverter::with_illuminants(
            Illuminant::D65,
            Illuminant::C,
            ChromaticAdaptation::XYZScaling
        )?;
        let c_result = c_converter.srgb_to_munsell(rgb)?;
        println!("  C:   {:.1}{} {:.1}/{:.1}", 
            c_result.hue, c_result.family, c_result.value, c_result.chroma);
        
        // Check if they differ
        let hue_diff = (d65_result.hue - c_result.hue).abs();
        let value_diff = (d65_result.value - c_result.value).abs();
        let chroma_diff = (d65_result.chroma - c_result.chroma).abs();
        
        if hue_diff > 0.1 || value_diff > 0.1 || chroma_diff > 0.1 {
            println!("  ✅ Illuminant adaptation is working (differences detected)");
        } else {
            println!("  ⚠️ No significant difference between illuminants");
        }
    }
    
    if exact_pct >= 55.0 && exact_pct <= 65.0 {
        println!("\n✅ SUCCESS: Accuracy is in expected range (~60.4%) for breakthrough version!");
    } else {
        println!("\n⚠️ WARNING: Accuracy {:.1}% is outside expected range (55-65%)", exact_pct);
    }
    
    Ok(())
}